zkp-prover/src/zkp_component.rs

```diff
@@ -0,0 +1,283 @@
+            zkp_requests: Arc::clone(&self.zkp_requests),
+            zkp_events: self.zkp_events.clone(),
+        }
+    }
+
+    pub fn is_zkp_prover_activated(&self) -> bool {
+        self.zk_prover.is_some()
+    }
+
+    fn load_proof_from_db(
+        blockchain: &Arc<RwLock<Blockchain>>,
+        zkp_state: &Arc<RwLock<ZKPState>>,
+        proof_storage: &ProofStore,
+        zkp_events: &BroadcastSender<ZKProof>,
+    ) {
+        // Load from db the preexisting proof if any.
+        if let Some(loaded_proof) = proof_storage.get_zkp() {
+            if let Err(e) = Self::push_proof_from_peers(
+                blockchain,
+                zkp_state,
+                loaded_proof,
+                &mut None,
+                proof_storage,
+                zkp_events,
+                false,
+            ) {
+                log::error!("Error pushing the zk proof load from disk {}", e);
+                proof_storage.set_zkp(&zkp_state.read().clone().into());
+            } else {
+                log::info!("The zk proof was successfully load from disk");
+            }
+        } else {
+            log::trace!("No zk proof found on the db");
+        }
+    }
+
+    fn push_proof_from_peers(
+        blockchain: &Arc<RwLock<Blockchain>>,
+        zkp_state: &Arc<RwLock<ZKPState>>,
+        zk_proof: ZKProof,
+        zk_prover: &mut Option<ZKProver<N>>,
+        proof_storage: &ProofStore,
+        zkp_events: &BroadcastSender<ZKProof>,
+        add_to_storage: bool,
+    ) -> Result<(), ZKPComponentError> {
+        let (new_block, genesis_block, proof) = pre_proof_validity_checks(blockchain, &zk_proof)?;
+        let zkp_state_lock = zkp_state.upgradable_read();
+
+        if new_block.block_number() <= zkp_state_lock.latest_block_number {
+            return Err(ZKPComponentError::OutdatedProof);
+        }
+        let new_zkp_state = validate_proof_get_new_state(proof, new_block, genesis_block)?;
+
+        let mut zkp_state_lock = RwLockUpgradableReadGuard::upgrade(zkp_state_lock);
+        *zkp_state_lock = new_zkp_state;
+
+        // Since someone else generate a valide proof faster, we will terminate our own proof generation.
+        if let Some(zk_prover) = zk_prover {
+            zk_prover.cancel_current_proof_production();
+        }
+
+        if add_to_storage {
+            proof_storage.set_zkp(&zkp_state_lock.clone().into())
+        }
+
+        if let Err(e) = zkp_events.send(zk_proof) {
+            log::error!("Error sending the proof to stream {}", e);
+        }
+
+        Ok(())
+    }
+}
+
+impl<N: Network> Future for ZKPComponent<N> {
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let this = self.project();
+
+        // Check if we have requested and are waiting for peers proofs
+        // Goes over all received proofs and tries to update state
+        // When  all requests were addressed and processed we reset the zkp_request variable
```

Comment:
```suggestion
        // Check if we have requested and are waiting for peers proofs.
        // Goes over all received proofs and tries to update state.
        // When all requests were addressed and processed we reset the `zkp_request` variable.
```

Reply:
Actually, I think the last sentence is not true.
```suggestion
        // Check if we have requested and are waiting for peers proofs.
        // Goes over all received proofs and tries to update state.
```